### PR TITLE
Unable to index the registerMainchain transaction

### LINKED ISSUE #1554: Unable to index the registerMainchain transaction
### Actual behavior

When indexing the `interoperability:registerMainchain` transaction, `getChainAccount` is invoked with the current chainID (sidechain), instead of the mainchain.

This results in the `Key ... not found.` error from `interoperablity_getChainAccount` endpoint.

### Expected behavior

The the indexing logic should be correct and the `interoperablity_getChainAccount` endpoint must be invoked with the mainchain chainID, to get the mainchain status. 

### Steps to reproduce

- Send an `interoperability:registerMainchain` transaction against a sidechain
- Run Lisk Service against it
- The indexing doesn't reach 100%

### Which version(s) does this affect? (Environment, OS, etc...)

Lisk Service v0.7.0-alpha.10

### Additional Information

- Revert the endpoint (`interoperablity_getChainAccount`) related changes made in https://github.com/LiskHQ/lisk-service/issues/1535 (PR https://github.com/LiskHQ/lisk-service/pull/1553) after the bug is fixed.
- Use the SDK endpoint to fetch the mainchainID and query the chainAccount Ref: https://github.com/LiskHQ/lisk-sdk/issues/8309
- Temporarily implement a similar logic (https://github.com/LiskHQ/lisk-sdk/blob/v6.0.0-alpha.16/framework/src/modules/interoperability/utils.ts#L386-L390) in the codebase (blockchain-connector) until the above issue is addressed
- Update the indexing logic for `services/blockchain-indexer/shared/indexer/transactionProcessor/interoperability/registerMainchain.js`
  - Fetch the mainchainID from the SDK via connector
  - Fetch the chainAccount by mainchainID and use the `status` and `name` to index
  - Use a similar approach for the revertTransaction hook as well

### PR DESCRIPTION
**Pull request by priojeetpriyom**
### What was the problem?
This PR resolves #1554

### How was it solved?
- [x] Revert the endpoint related changes made in https://github.com/LiskHQ/lisk-service/issues/1535.
- [x] Fetch the mainchainID from the SDK via connector. Temporarily implement a similar logic (https://github.com/LiskHQ/lisk-sdk/blob/v6.0.0-alpha.16/framework/src/modules/interoperability/utils.ts#L386-L390)
- [x] Update the indexing logic for services/blockchain-indexer/shared/indexer/transactionProcessor/interoperability/registerMainchain.js
- [x] Fetch the chainAccount by mainchainID and use the status and name to index
- [x] Use a similar approach for the revertTransaction hook as well
### How was it tested?
- [x] Local

### CODE REVIEW

File: services/blockchain-connector/shared/sdk/interoperability.js [hunk 1]
```diff
@@ -46,6 +35,24 @@ const getChainAccount = async (chainID) => {
...
+		// const mainchainID = await invokeEndpoint('interoperability_getMainchainID', { chainID });
+		return mainchainID;
+	} catch (err) {
+		if (err.message.includes(timeoutMessage)) {
+			throw new TimeoutException('Request timed out when calling \'getMainchainID\'.');
+		}
+		logger.warn(`Error returned when invoking 'interoperability_getMainchainID' with chainID: ${chainID}.\n${err.stack}`);
+		throw err;
+	}
+};
```
Comment (sameersubudhi) [inline services/blockchain-connector/shared/sdk/interoperability.js:53]: This must be implemented in a similar fashion below:
Also, please cache the `mainchainID`.

```
const getMainchainID = async () => {
	try {
		// const mainchainID = await invokeEndpoint('interoperability_getMainchainID', { chainID });

		// TODO: Remove this and use SDK endpoint once following issue is closed: https://github.com/LiskHQ/lisk-sdk/issues/8309
		const { chainID } = await getNodeInfo();
		const LENGTH_CHAIN_ID = 4 * 2; // Each byte is represented with 2 nibbles
		const mainchainID = chainID.substring(0, 2).padEnd(LENGTH_CHAIN_ID, '0');

		return mainchainID;
	} catch (err) {
		if (err.message.includes(timeoutMessage)) {
			throw new TimeoutException('Request timed out when calling \'getMainchainID\'.');
		}
		logger.warn(`Error returned when invoking 'interoperability_getMainchainID' with chainID: ${chainID}.\n${err.stack}`);
		throw err;
	}
};
```

File: services/blockchain-connector/shared/sdk/pos.js [hunk 2]
```diff
@@ -94,14 +93,6 @@ const getPosPendingUnlocks = async (address) => {
...
 	try {
 		const staker = await invokeEndpoint('pos_getStaker', { address });
-
-		if (staker.error && regex.KEY_NOT_EXIST.test(staker.error.message)) {
-			return {
-				stakes: [],
-				pendingUnlocks: [],
-			};
-		}
-
```
Comment (sameersubudhi) [inline services/blockchain-connector/shared/sdk/pos.js:104]: Please revert this change.

File: services/blockchain-connector/shared/utils/regex.js [hunk 3]
```diff
@@ -16,11 +16,9 @@
 const BLS_KEY = /^\b[a-fA-F0-9]{96}\b$/;
 const CHAIN_ID = /^\b[a-fA-F0-9]{8}\b$/;
 const PROOF_OF_POSSESSION = /^\b[a-fA-F0-9]{192}\b$/;
-const KEY_NOT_EXIST = /^Key [a-f0-9]+ does not exist\.$/;
```
Comment (sameersubudhi) [inline services/blockchain-connector/shared/utils/regex.js:19]: Please revert.

File: services/blockchain-indexer/shared/dataService/business/interoperability/mainchain.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+ * Removal or modification of this copyright notice is prohibited.
+ *
+ */
+
+const { requestConnector } = require('../../../utils/request');
+
+const getMainchainID = async params => {
+	const response = await requestConnector('getMainchainID', { chainID: params.chainID });
+	return response;
+};
```
Comment (sameersubudhi) [inline services/blockchain-indexer/shared/dataService/business/interoperability/mainchain.js:22]:
```suggestion
const getMainchainID = async() => {
	const response = await requestConnector('getMainchainID');
	return response;
};
```

File: services/blockchain-indexer/shared/dataService/interoperability/mainchain.js [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ *
+ */
+const business = require('../business');
+
+const getMainchainID = async params => {
+	const response = await business.getMainchainID(params);
+	return response;
+};
```
Comment (sameersubudhi) [inline services/blockchain-indexer/shared/dataService/interoperability/mainchain.js:21]:
```suggestion
const getMainchainID = async () => business.getMainchainID();
```

File: services/blockchain-indexer/shared/indexer/transactionProcessor/interoperability/registerMainchain.js [hunk 6]
```diff
@@ -39,46 +39,52 @@ const getBlockchainAppsTable = () => getTableInstance(
...
+		chainStatus,
+	};
 };
 
 const applyTransaction = async (blockHeader, tx, events, dbTrx) => {
 	if (tx.executionStatus !== TRANSACTION_STATUS.SUCCESS) return;
 
 	const blockchainAppsTable = await getBlockchainAppsTable();
-	const chainStatus = await getChainStatus(tx.params.ownChainID);
+	const mainchainID = await getMainchainID({ chainID: tx.params.ownChainID });
```
Comment (sameersubudhi) [inline services/blockchain-indexer/shared/indexer/transactionProcessor/interoperability/registerMainchain.js:56]:
```suggestion
	const mainchainID = await getMainchainID();
```

File: services/blockchain-indexer/shared/indexer/transactionProcessor/interoperability/registerMainchain.js [hunk 7]
```diff
@@ -39,46 +39,52 @@ const getBlockchainAppsTable = () => getTableInstance(
...
+	const { chainName, chainStatus } = await getChainInfo(mainchainID);
```
Comment (sameersubudhi) [inline services/blockchain-indexer/shared/indexer/transactionProcessor/interoperability/registerMainchain.js:57]:
```suggestion
	const mainchainInfo = await getChainInfo(mainchainID);
```

File: services/blockchain-indexer/shared/indexer/transactionProcessor/interoperability/registerMainchain.js [hunk 8]
```diff
@@ -39,46 +39,52 @@ const getBlockchainAppsTable = () => getTableInstance(
...
-const getChainStatus = async chainID => {
-	const { status: chainStatusInt } = await getChainAccount({ chainID });
+const getChainInfo = async chainID => {
+	const chainAccount = await getChainAccount({ chainID });
+	const chainStatusInt = chainAccount.status;
 	const chainStatus = CHAIN_STATUS[chainStatusInt];
-	return chainStatus;
+	return {
+		chainName: chainAccount.name,
+		chainStatus,
```
Comment (sameersubudhi) [inline services/blockchain-indexer/shared/indexer/transactionProcessor/interoperability/registerMainchain.js:48]:
```suggestion
		name: chainAccount.name,
		status,
```

File: services/blockchain-indexer/shared/indexer/transactionProcessor/interoperability/registerMainchain.js [hunk 9]
```diff
@@ -39,46 +39,52 @@ const getBlockchainAppsTable = () => getTableInstance(
...
+	const { chainName, chainStatus } = await getChainInfo(mainchainID);
 
-	logger.trace(`Indexing mainchain (${tx.params.chainID}) registration information.`);
+	logger.trace(`Indexing mainchain (${mainchainID}) registration information.`);
 	const appInfo = {
-		chainID: tx.params.ownChainID,
-		name: tx.params.ownName,
+		chainID: mainchainID,
+		name: chainName,
 		status: chainStatus,
```
Comment (sameersubudhi) [inline services/blockchain-indexer/shared/indexer/transactionProcessor/interoperability/registerMainchain.js:63]:
```suggestion
		name: mainchainInfo.name,
		status: mainchainInfo.status,
```

File: services/blockchain-indexer/shared/indexer/transactionProcessor/interoperability/registerMainchain.js [hunk 10]
```diff
@@ -39,46 +39,52 @@ const getBlockchainAppsTable = () => getTableInstance(
...
-	logger.debug(`Indexed mainchain (${tx.params.chainID}) registration information.`);
+	logger.debug(`Indexed mainchain (${mainchainID}) registration information.`);
 };
 
 // eslint-disable-next-line no-unused-vars
 const revertTransaction = async (blockHeader, tx, events, dbTrx) => {
 	if (tx.executionStatus !== TRANSACTION_STATUS.SUCCESS) return;
 
 	const blockchainAppsTable = await getBlockchainAppsTable();
+	const mainchainID = await getMainchainID({ chainID: tx.params.ownChainID });
```
Comment (sameersubudhi) [inline services/blockchain-indexer/shared/indexer/transactionProcessor/interoperability/registerMainchain.js:78]:
```suggestion
	const mainchainID = await getMainchainID();
```

File: services/blockchain-indexer/shared/indexer/transactionProcessor/interoperability/registerSidechain.js [hunk 11]
```diff
@@ -27,7 +27,7 @@ const logger = Logger();
 const MYSQL_ENDPOINT = config.endpoints.mysql;
 const blockchainAppsTableSchema = require('../../../database/schema/blockchainApps');
 const { TRANSACTION_STATUS } = require('../../../constants');
-const { getChainStatus } = require('./registerMainchain');
+const { getChainInfo } = require('./registerMainchain');
```
Comment (sameersubudhi) [inline services/blockchain-indexer/shared/indexer/transactionProcessor/interoperability/registerSidechain.js:30]: Refactor accordingly.

File: services/blockchain-indexer/shared/indexer/transactionProcessor/interoperability/submitSidechainCrossChainUpdate.js [hunk 12]
```diff
@@ -40,7 +40,7 @@ const applyTransaction = async (blockHeader, tx, events, dbTrx) => {
 	if (tx.executionStatus !== TRANSACTION_STATUS.SUCCESS) return;
 
 	const blockchainAppsTable = await getBlockchainAppsTable();
-	const chainStatus = await getChainStatus(tx.params.sendingChainID);
+	const { chainStatus } = await getChainInfo(tx.params.sendingChainID);
```
Comment (sameersubudhi) [inline services/blockchain-indexer/shared/indexer/transactionProcessor/interoperability/submitSidechainCrossChainUpdate.js:43]: Refactor accordingly.

File: services/blockchain-indexer/shared/indexer/transactionProcessor/interoperability/terminateSidechainForLiveness.js [hunk 13]
```diff
@@ -40,7 +40,7 @@ const applyTransaction = async (blockHeader, tx, events, dbTrx) => {
 	if (tx.executionStatus !== TRANSACTION_STATUS.SUCCESS) return;
 
 	const blockchainAppsTable = await getBlockchainAppsTable();
-	const chainStatus = await getChainStatus(tx.params.chainID);
+	const { chainStatus } = await getChainInfo(tx.params.chainID);
```
Comment (sameersubudhi) [inline services/blockchain-indexer/shared/indexer/transactionProcessor/interoperability/terminateSidechainForLiveness.js:43]: Refactor accordingly.

**Review by sameersubudhi — CHANGES_REQUESTED** [review]

File: services/blockchain-indexer/shared/indexer/transactionProcessor/interoperability/terminateSidechainForLiveness.js (hunk 13 above)
Comment (priojeetpriyom) [reply services/blockchain-indexer/shared/indexer/transactionProcessor/interoperability/terminateSidechainForLiveness.js:43]: Applied.

File: services/blockchain-indexer/shared/indexer/transactionProcessor/interoperability/submitSidechainCrossChainUpdate.js (hunk 12 above)
Comment (priojeetpriyom) [reply services/blockchain-indexer/shared/indexer/transactionProcessor/interoperability/submitSidechainCrossChainUpdate.js:43]: Applied.

File: services/blockchain-indexer/shared/indexer/transactionProcessor/interoperability/registerMainchain.js (hunk 10 above)
Comment (priojeetpriyom) [reply services/blockchain-indexer/shared/indexer/transactionProcessor/interoperability/registerMainchain.js]: Applied.

File: services/blockchain-indexer/shared/indexer/transactionProcessor/interoperability/registerMainchain.js (hunk 9 above)
Comment (priojeetpriyom) [reply services/blockchain-indexer/shared/indexer/transactionProcessor/interoperability/registerMainchain.js]: Applied.

File: services/blockchain-indexer/shared/indexer/transactionProcessor/interoperability/registerMainchain.js (hunk 7 above)
Comment (priojeetpriyom) [reply services/blockchain-indexer/shared/indexer/transactionProcessor/interoperability/registerMainchain.js]: Applied.

File: services/blockchain-indexer/shared/indexer/transactionProcessor/interoperability/registerMainchain.js (hunk 6 above)
Comment (priojeetpriyom) [reply services/blockchain-indexer/shared/indexer/transactionProcessor/interoperability/registerMainchain.js]: Applied.

File: services/blockchain-indexer/shared/dataService/interoperability/mainchain.js (hunk 5 above)
Comment (priojeetpriyom) [reply services/blockchain-indexer/shared/dataService/interoperability/mainchain.js]: Applied.

File: services/blockchain-indexer/shared/dataService/business/interoperability/mainchain.js (hunk 4 above)
Comment (priojeetpriyom) [reply services/blockchain-indexer/shared/dataService/business/interoperability/mainchain.js]: Applied.

File: services/blockchain-connector/shared/utils/regex.js (hunk 3 above)
Comment (priojeetpriyom) [reply services/blockchain-connector/shared/utils/regex.js]: Applied.

File: services/blockchain-connector/shared/sdk/pos.js (hunk 2 above)
Comment (priojeetpriyom) [reply services/blockchain-connector/shared/sdk/pos.js]: Applied.

File: services/blockchain-connector/shared/sdk/interoperability.js (hunk 1 above)
Comment (priojeetpriyom) [reply services/blockchain-connector/shared/sdk/interoperability.js]: Applied

File: services/blockchain-connector/shared/sdk/interoperability.js [hunk 14]
```diff
@@ -46,6 +36,26 @@ const getChainAccount = async (chainID) => {
 	}
 };
 
+const getMainchainID = async () => {
+	try {
+		// const mainchainID = await invokeEndpoint('interoperability_getMainchainID', { chainID });
```
Comment (nagdahimanshu) [inline services/blockchain-connector/shared/sdk/interoperability.js:41]: If given method `interoperability_getMainchainID` requires `chainID` parameter as defined in L41, then we should add chainID as input parameter to `getMainchainID`, please update accordingly.
```suggestion
const getMainchainID = async (chainID) => {
	try {
		// const mainchainID = await invokeEndpoint('interoperability_getMainchainID', { chainID });
```

Comment (sameersubudhi) [reply services/blockchain-connector/shared/sdk/interoperability.js:41]: @nagdahimanshu  The method signature is not yet defined but I expect no params to be passed here.

@priojeetpriyom Please cache the mainchainID here, as requested previously.

```suggestion
const getMainchainID = async () => {
	try {
		// const mainchainID = await invokeEndpoint('interoperability_getMainchainID');
```

**Review by sameersubudhi — CHANGES_REQUESTED** [review]
Looks good overall except for the following.

File: services/blockchain-connector/shared/sdk/interoperability.js (hunk 14 above)
Comment (priojeetpriyom) [reply services/blockchain-connector/shared/sdk/interoperability.js]: Applied.

**Review by sameersubudhi — APPROVED** [review]

**Review by nagdahimanshu — APPROVED** [review]
Looks good 👍

**Pull request merged**